File: src/Reloaded.Memory.Buffers/Structs/Params/BufferAllocatorSettings.cs
using System.Diagnostics;
using JetBrains.Annotations;
using Reloaded.Memory.Buffers.Utilities;

namespace Reloaded.Memory.Buffers.Structs.Params;

/// <summary>
///     Settings to pass to the buffer allocator.
/// </summary>
[PublicAPI]
public struct BufferAllocatorSettings
{
    /// <summary>
    ///     Minimum address of the allocation.
    /// </summary>
    public required nuint MinAddress { get; set; } = 0;

    /// <summary>
    ///     Maximum address of the allocation.
    /// </summary>
    public required nuint MaxAddress { get; set; } = Cached.GetMaxAddress();

    /// <summary>
    ///     Required size of the data.
    /// </summary>
    public required uint Size { get; set; } = 4096;

    /// <summary>
    ///     Process to allocate memory in.
    /// </summary>
    public Process TargetProcess { get; init; } = Cached.GetThisProcess();

    /// <summary>
    ///     Amount of times library should retry after failing to allocate a region.
    /// </summary>
    /// <remarks>
    ///     This is useful when there's high memory pressure, meaning pages become unavailable between the time
    ///     they are found and the time we try to allocate them.
    /// </remarks>
    public int RetryCount { get; set; } = 8;

    /// <summary>
    ///     Whether to use brute force to find a suitable address.
    /// </summary>
    /// <remarks>
    ///     This for some reason only ever was needed in FFXIV under Wine; and was contributed in the original library
    ///     (prior to rewrite) by the Dalamud folks. In Wine and on FFXIV *only*; the regular procedure of trying to allocate
    ///     returned pages doesn't always work. This is a last ditch workaround for that.<br />
    ///     This setting is only used on Windows targets today.
    /// </remarks>
    public bool BruteForce { get; set; } = true;

    /// <summary>
    ///     Initializes the buffer allocator with default settings.
    /// </summary>
    public BufferAllocatorSettings() { }

    /// <summary>
    ///     Creates settings such that the returned buffer will always be within <paramref name="proximity"/> bytes of <paramref name="target"/>.
    /// </summary>
    /// <param name="proximity">Max proximity (number of bytes) to target.</param>
    /// <param name="target">Target address.</param>
    /// <param name="size">Size required in the settings.</param>
    /// <returns>Settings that would satisfy this search.</returns>
    public static BufferAllocatorSettings FromProximity(nuint proximity, nuint target, nuint size)
    {
        return new BufferAllocatorSettings()
        {
            MaxAddress = Mathematics.AddWithOverflowCap(target, proximity),
            MinAddress = Mathematics.SubtractWithUnderflowCap(target, proximity),
            Size = (uint)size
        };
    }

    /// <summary>
    ///     Sanitizes the input values.
    /// </summary>
    public void Sanitize()
    {
        // On Windows, VirtualAlloc treats 0 as 'any address', we might aswell avoid this out the gate.
        if (Polyfills.IsWindows() && MinAddress < (ulong)Cached.GetAllocationGranularity())
            MinAddress = (nuint)Cached.GetAllocationGranularity();

        Size = (uint)Mathematics.RoundUp(Size, Cached.GetAllocationGranularity());
    }
}

File: src-rust/src/structs/params/buffer_allocator_settings.rs
use crate::utilities::cached::CACHED;
use crate::utilities::mathematics;
use std::cmp;

/// Settings to pass to the buffer allocator.
#[derive(Debug, Clone, Copy)]
#[repr(C)]
pub struct BufferAllocatorSettings {
    /// Minimum address of the allocation.
    pub min_address: usize,

    /// Maximum address of the allocation.
    pub max_address: usize,

    /// Required size of the data.
    pub size: u32,

    /// Process to allocate memory in.
    /// Stored as process id.
    pub target_process_id: u32,

    /// Amount of times library should retry after failing to allocate a region.
    ///
    /// # Remarks
    /// This is useful when there's high memory pressure, meaning pages become unavailable between the time
    /// they are found and the time we try to allocate them.
    pub retry_count: i32,

    /// Whether to use brute force to find a suitable address.
    ///
    /// # Remarks
    ///
    /// In the original library, this was for some reason only ever needed in FFXIV under Wine; and was contributed
    /// (prior to rewrite) by the Dalamud folks. In Wine and on FFXIV *only*; this was ever the case.
    /// Inclusion of a brute force approach is a last ditch workaround for that.
    ///
    /// This setting is only used on Windows targets today.
    pub brute_force: bool,
}

impl BufferAllocatorSettings {
    /// Initializes the buffer allocator with default settings.
    pub fn new() -> Self {
        Self {
            min_address: 0,
            max_address: CACHED.max_address,
            size: 4096,
            target_process_id: CACHED.this_process_id,
            retry_count: 8,
            brute_force: true,
        }
    }

    /// Creates settings such that the returned buffer will always be within `proximity` bytes of `target`.
    ///
    /// # Arguments
    ///
    /// * `proximity` - Max proximity (number of bytes) to target.
    /// * `target` - Target address.
    /// * `size` - Size required in the settings.
    ///
    /// # Returns
    ///
    /// * `BufferAllocatorSettings` - Settings that would satisfy this search.
    pub fn from_proximity(proximity: usize, target: usize, size: usize) -> Self {
        Self {
            max_address: mathematics::add_with_overflow_cap(target, proximity),
            min_address: mathematics::subtract_with_underflow_cap(target, proximity),
            size: size as u32,
            ..Self::new()
        }
    }

    /// Sanitizes the input values.
    pub fn sanitize(&mut self) {
        // On Windows, VirtualAlloc treats 0 as 'any address', we might aswell avoid this out the gate.
        if cfg!(windows) && (self.min_address < CACHED.get_allocation_granularity() as usize) {
            self.min_address = CACHED.get_allocation_granularity() as usize;
        }

        self.size = cmp::max(self.size, 1);
        self.size = mathematics::round_up(
            self.size as usize,
            CACHED.get_allocation_granularity() as usize,
        ) as u32;
    }
}

impl Default for BufferAllocatorSettings {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test_default_settings() {
        let settings = BufferAllocatorSettings::new();
        assert_eq!(settings.min_address, 0);
        assert_eq!(settings.max_address, CACHED.max_address);
        assert_eq!(settings.size, 4096);
        assert_eq!(settings.target_process_id, CACHED.this_process_id);
        assert_eq!(settings.retry_count, 8);
        assert!(settings.brute_force);
    }

    #[test]
    fn test_from_proximity() {
        let proximity: usize = 1000;
        let target: usize = 2000;
        let size: usize = 3000;
        let settings = BufferAllocatorSettings::from_proximity(proximity, target, size);

        assert_eq!(
            settings.max_address,
            mathematics::add_with_overflow_cap(target, proximity)
        );
        assert_eq!(
            settings.min_address,
            mathematics::subtract_with_underflow_cap(target, proximity)
        );
        assert_eq!(settings.size, size as u32);
        assert_eq!(settings.target_process_id, CACHED.this_process_id);
        assert_eq!(settings.retry_count, 8);
        assert!(settings.brute_force);
    }

    #[test]
    fn test_sanitize() {
        let mut settings = BufferAllocatorSettings::new();
        settings.min_address = 0;
        settings.size = 1;

        settings.sanitize();

        if cfg!(windows) {
            assert_eq!(
                settings.min_address,
                CACHED.get_allocation_granularity() as usize
            );
        } else {
            assert_eq!(settings.min_address, 0);
        }

        assert_eq!(
            settings.size,
            mathematics::round_up(
                cmp::max(1, 1) as usize,
                CACHED.get_allocation_granularity() as usize
            ) as u32
        );
    }
}
